R/unnest-longer.R

```diff
@@ -141,97 +149,89 @@ col_to_long <- function(col,
-    indices_include = indices_include,
-    error_call = error_call
-  )
-  elt_ptype <- vec_ptype(elt_ptype)
-
-  # Avoid expensive dispatch from `[[.list_of`, and allow for `[[<-`
+  # Avoid expensive dispatch from `[[.list_of`, and allow for `[[<-`.
+  # We've already captured the `ptype`.
   col <- tidyr_new_list(col)
 
-  for (i in seq_along(col)) {
-    col[[i]] <- elt_to_long(
-      x = col[[i]],
-      index = indices[[i]],
-      name = name,
-      values_to = values_to,
-      indices_to = indices_to,
-      indices_include = indices_include,
-      error_call = error_call
+  if (!list_all_vectors2(col)) {
+    cli::cli_abort(
+      "List-column {.var {name}} must contain only vectors or `NULL`.",
+      call = error_call
     )
   }
 
-  ptype <- vec_ptype_common(elt_ptype, !!!col)
-  col <- vec_cast_common(!!!col, .to = ptype)
+  sizes <- list_sizes(col)
 
-  col <- new_list_of(col, ptype = ptype)
+  # Collect index info before replacing `NULL`s so `keep_empty` works correctly
+  info <- collect_indices_info(col, sizes, indices_include, keep_empty)
+  indices_include <- info$indices_include
+  indices <- info$indices
+  index_ptype <- info$index_ptype
 
-  col
-}
+  size_null <- as.integer(keep_empty)
 
-# Convert a list element to a long tibble with:
-# - 1 col (2 if `indices_include` is `TRUE`)
-# - N rows, where `N = vec_size(x)`
-elt_to_long <- function(x,
-                        index,
-                        name,
-                        values_to,
-                        indices_to,
-                        indices_include,
-                        error_call = caller_env()) {
-  if (is.null(x)) {
-    x <- unspecified(1L)
-
-    if (indices_include) {
-      if (is.integer(index)) {
-        index <- 1L
-      } else {
-        index <- NA_character_
-      }
-    }
+  info <- list_init_null(col, sizes, ptype = ptype, size = size_null)
+  col <- info$x
+  sizes <- info$sizes
+
+  if (keep_empty) {
+    info <- list_init_typed(col, sizes, ptype = ptype, size = 1L)
+    col <- info$x
+    sizes <- info$sizes
   }
 
-  if (!vec_is(x)) {
-    cli::cli_abort(
-      "List-column {.var {name}} must contain only vectors.",
-      call = error_call
-    )
+  if (is.null(ptype)) {
+    # Initialize `ptype` to generate a `ptype` version of the output data frame.
+    # Important in the size 0 input case.
+    ptype <- unspecified()
   }
 
   if (indices_include) {
-    out <- list(x, index)
-    names(out) <- c(values_to, indices_to)
+    names <- c(values_to, indices_to)
+    ptype <- new_long_indexed_frame(ptype, index_ptype, 0L, names)
+    col <- pmap(
+      list(col, indices, sizes),
+      function(elt, index, size) new_long_indexed_frame(elt, index, size, names)
+    )
   } else {
-    out <- list(x)
-    names(out) <- values_to
+    name <- values_to
+    ptype <- new_long_frame(ptype, 0L, name)
+    col <- map2(
+      col,
+      sizes,
+      function(elt, size) new_long_frame(elt, size, name)
+    )
```

Comment:
The `map2()` and `pmap()` paths are now the most expensive part of `unnest_longer()`. Even though the functions are fairly simple (wrap into a data frame), if `col` is long the repeated calls to `.f` just naturally take awhile.